src/playercards/AllCardsBag.ttslua

```diff
@@ -147,89 +333,216 @@ function buildSupplementalIndexes()
-      local isGuardian = false
-      local isSeeker = false
-      local isMystic = false
-      local isRogue = false
-      local isSurvivor = false
-      local isNeutral = false
-      local upgradeKey
-      -- Excludes signature cards (which have no class or level) and alternate
-      -- ID entries
-      if (cardMetadata.class ~= nil and cardMetadata.level ~= nil) then
-        isGuardian = string.match(cardMetadata.class, "Guardian")
-        isSeeker = string.match(cardMetadata.class, "Seeker")
-        isMystic = string.match(cardMetadata.class, "Mystic")
-        isRogue = string.match(cardMetadata.class, "Rogue")
-        isSurvivor = string.match(cardMetadata.class, "Survivor")
-        isNeutral = string.match(cardMetadata.class, "Neutral")
-        if (cardMetadata.level > 0) then
-          upgradeKey = "-upgrade"
-        else
-          upgradeKey = "-level0"
+      table.insert(allCardsIndex, cardMetadata.id)
+
+      if cardMetadata.level then
+        table.insert(levelIndex[cardMetadata.level], cardMetadata.id)
+      else
+        otherIndex[cardMetadata.id] = true
+      end
+
+      if cardMetadata.class then
+        local classString = cardMetadata.class .. "|"
+        for class in classString:gmatch("[^|]+") do
+          table.insert(classIndex[string.lower(class)], cardMetadata.id)
         end
-        if (isGuardian) then
-          table.insert(classAndLevelIndex["Guardian"..upgradeKey], cardMetadata.id)
+        local classCount = select(2, string.gsub(classString, "|", ""))
+        table.insert(classIndex[classCount], cardMetadata.id)
+      end    
+
+      local cycleName = cardMetadata.cycle
+      if cycleName ~= nil then
+        cycleName = string.lower(cycleName)
+        if string.match(cycleName, "return") then
+          cycleName = string.sub(cycleName, 11)
         end
-        if (isSeeker) then
-          table.insert(classAndLevelIndex["Seeker"..upgradeKey], cardMetadata.id)
+        if cycleName == "the night of the zealot" then
+          cycleName = "core"
         end
-        if (isMystic) then
-          table.insert(classAndLevelIndex["Mystic"..upgradeKey], cardMetadata.id)
+        cycleIndex[cycleName] = cycleIndex[cycleName] or {}
+        table.insert(cycleIndex[cycleName], cardMetadata.id)
+      end
+
+      local bonded = cardMetadata.bonded
+
+      local checkedTraits = {}
+      if cardMetadata.traits then
+        for traitString in string.gmatch(cardMetadata.traits, "%w+.") do
+          local trait = string.lower(string.gsub(traitString, "[.]", ""))
+          checkedTraits[trait] = true
+          if traitsIndex[trait] == nil then
+            traitsIndex[trait] = { }
+          end
+          table.insert(traitsIndex[trait], cardMetadata.id)
         end
-        if (isRogue) then
-          table.insert(classAndLevelIndex["Rogue"..upgradeKey], cardMetadata.id)
+      end
+
+      if cardMetadata.willpowerIcons then
```

Comment:
Instead of repeating the same code 4 times, you could use a loop for this:
```lua
for _, symbol in ipairs({"combat", "intellect", "willpower", "agility", "wild") do
  local index = _G[symbol  .. "IconsIndex"]
  local iconCount = cardMetadata[symbol .. "Icons"] or 0
  index[iconCount] = index[iconCount] or { }
  table.insert(index[iconCount ], cardMetadata.id)
end
```

Reply:
Ooo very smart to inline the ipairs call!  Making this change now!